state-chain/runtime/src/chainflip/simulate_swap.rs

```diff
@@ -0,0 +1,232 @@
+// Copyright 2025 Chainflip Labs GmbH
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+// SPDX-License-Identifier: Apache-2.0
+
+use cf_chains::{
+	assets::any::ForeignChainAndAsset,
+	instances::{
+		ArbitrumInstance, AssethubInstance, BitcoinInstance, EthereumInstance, PolkadotInstance,
+		SolanaInstance,
+	},
+};
+use cf_primitives::{
+	AccountId, Asset, AssetAmount, BasisPoints, Beneficiary, DcaParameters, IngressOrEgress,
+};
+use cf_traits::OrderId;
+use pallet_cf_swapping::{BatchExecutionError, FeeType, NetworkFeeTracker, Swap};
+use sp_runtime::DispatchError;
+
+use crate::{
+	runtime_apis::{
+		CcmData, DispatchErrorWithMessage, FeeTypes, SimulateSwapAdditionalOrder,
+		SimulatedSwapInformation,
+	},
+	LiquidityPools, Runtime, Swapping,
+};
+
+use sp_runtime::traits::UniqueSaturatedInto;
+use sp_std::{collections::btree_set::BTreeSet, vec, vec::Vec};
+
+pub fn simulate_swap(
+	input_asset: Asset,
+	output_asset: Asset,
+	input_amount: AssetAmount,
+	broker_commission: BasisPoints,
+	dca_parameters: Option<DcaParameters>,
+	ccm_data: Option<CcmData>,
+	exclude_fees: BTreeSet<FeeTypes>,
+	additional_orders: Option<Vec<SimulateSwapAdditionalOrder>>,
+	is_internal: Option<bool>,
+) -> Result<SimulatedSwapInformation, DispatchErrorWithMessage> {
+	let is_internal = is_internal.unwrap_or_default();
+	let mut exclude_fees = exclude_fees;
+	if is_internal {
+		exclude_fees.insert(FeeTypes::IngressDepositChannel);
+		exclude_fees.insert(FeeTypes::Egress);
+		exclude_fees.insert(FeeTypes::IngressVaultSwap);
+	}
+
+	if let Some(additional_orders) = additional_orders {
+		for (index, additional_order) in additional_orders.into_iter().enumerate() {
+			match additional_order {
+				SimulateSwapAdditionalOrder::LimitOrder {
+					base_asset,
+					quote_asset,
+					side,
+					tick,
+					sell_amount,
+				} => {
+					LiquidityPools::try_add_limit_order(
+						&AccountId::new([0; 32]),
+						base_asset,
+						quote_asset,
+						side,
+						index as OrderId,
+						tick,
+						sell_amount.into(),
+					)?;
+				},
+			}
+		}
+	}
+
+	fn remove_fees(
```

Comment:
Nit: there is no reason to nest this function inside `simulate_swap`. You already have a new scope since you created this new file, and so you won't be polluting external scope as long as the function is private.